### PR TITLE
Lazy loading & hot reloading: allow to inject & override api defintions on existing api

### PR DESCRIPTION
**Pull request by phryneas**

### CODE REVIEW

File: examples/svelte/src/App.svelte [hunk 1]
```diff
@@ -2,23 +2,22 @@
...
-    
+
     onMount(async () => {
-        queryRef = ({ refetch: getCount } = store.dispatch(counterApi.queryActions.getCount()));
+        queryRef = { refetch: getCount } = store.dispatch(actions.getCount());
     });
-    
-    $: ({ data } = counterApi.selectors.query.getCount()($store));
 
+    $: ({ data } = counterApi.selectors.getCount()($store));
```
Comment (msutkowski) [inline examples/svelte/src/App.svelte:20]:
```suggestion
    $: ({ data } = selectors.getCount()($store));
```

File: src/buildActionMaps.ts [hunk 2]
```diff
@@ -62,29 +76,41 @@ export type MutationActionCreatorResult<D extends MutationDefinition<any, any, a
...
+  function buildAction<D extends MutationDefinition<any, any, any, any>>(
+    endpoint: string,
+    definition: D
+  ): StartMutationActionCreator<D>;
+
+  function buildAction(endpoint: string, definition: EndpointDefinition<any, any, any, any>) {
+    if (isQueryDefinition(definition)) {
+      return buildQueryAction(endpoint, definition);
+    } else if (isMutationDefinition(definition)) {
+      return buildMutationAction(endpoint, definition);
```
Comment (msutkowski) [inline src/buildActionMaps.ts:109]: Readability wins 🏆

**Comment by markerikson** [comment]
yayyyy!

File: src/index.ts [hunk 3]
```diff
@@ -39,71 +46,110 @@ export function createApi<
...
+        api.selectors[endpoint] = buildMutationSelector(endpoint, definition);
+        api.actions[endpoint] = buildMutationAction(endpoint, definition);
+        api.hooks[endpoint] = { useMutation: buildMutationHook(endpoint) };
+      }
+    }
+
+    return api as any;
+  }
+
+  return api.injectEndpoints({ endpoints });
```
Comment (phryneas) [inline src/index.ts:135]: it's so :fire:  that this works

**Comment by phryneas** [comment]
Aaaand it's done. The API changed a bit and now has this shape:
```ts
export interface Api<
  InternalQueryArgs,
  Definitions extends EndpointDefinitions,
  ReducerPath extends string,
  EntityTypes extends string
> {
  reducerPath: ReducerPath;
  actions: EndpointActions<Definitions>;
  internalActions: SliceActions;
  reducer: Reducer<CombinedState<Definitions, EntityTypes> & QueryStatePhantomType<ReducerPath>, AnyAction>;
  selectors: Selectors<Definitions, RootState<Definitions, EntityTypes, ReducerPath>>;
  middleware: Middleware<{}, RootState<Definitions, string, ReducerPath>, ThunkDispatch<any, any, AnyAction>>;
  hooks: Hooks<Definitions>;
  injectEndpoints<NewDefinitions extends EndpointDefinitions>(_: {
    endpoints: (build: EndpointBuilder<InternalQueryArgs, EntityTypes>) => NewDefinitions;
    overrideExisting?: boolean;
  }): Api<InternalQueryArgs, Definitions & NewDefinitions, ReducerPath, EntityTypes>;
}
```

Also, there's the new `ApiWithInjectedEndpoints` type - we can still rename that if someone has a better idea.

File: examples/react/src/features/bundleSplitting/Post.tsx [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+const Post = ({ id }: { id: number }) => {
+  /*
+   * As accessing these conditionally will cause hooks to error out,
+   * you best assert their existence & throw otherwise.
+   * This missing would be a programming error that you should
+   * catch early anyways.
+   */
+  assert(splitApi.hooks.getPost, 'Endpoint `getPost` not loaded!');
+  const { data } = splitApi.hooks.getPost.useQuery(id);
+  return !data ? <>loading</> : <h1>{data.name}</h1>;
```
Comment (msutkowski) [inline examples/react/src/features/bundleSplitting/Post.tsx:29]:
```suggestion
  return error ? <>there was an error</> : !data ? <>loading</> : <h1>{data.name}</h1>;
```

File: examples/react/src/features/bundleSplitting/PostsList.tsx [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+   * That way all the injected endpoint are typed on there,
+   * but no endpoints that were injected elsewhere will be typed.
+   *
+   * They *will* be available at runtime if they have been
+   * injected though.
+   */
+
+  const { data } = apiWithPosts.hooks.getPosts.useQuery();
+  const [selected, select] = useState<number | undefined>();
+  return !data ? (
```
Comment (msutkowski) [inline examples/react/src/features/bundleSplitting/PostsList.tsx:18]:
```suggestion
  return error ? <>there was an error</> : !data ? (
```

File: examples/react/src/features/bundleSplitting/PostsList.tsx [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+  /**
+   * You can directly use the "injected" API.
+   * That way all the injected endpoint are typed on there,
+   * but no endpoints that were injected elsewhere will be typed.
+   *
+   * They *will* be available at runtime if they have been
+   * injected though.
+   */
+
+  const { data } = apiWithPosts.hooks.getPosts.useQuery();
```
Comment (msutkowski) [inline examples/react/src/features/bundleSplitting/PostsList.tsx:16]:
```suggestion
  const { data, error } = apiWithPosts.hooks.getPosts.useQuery();
```

File: examples/react/src/features/bundleSplitting/Post.tsx [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+
+const Post = ({ id }: { id: number }) => {
+  /*
+   * As accessing these conditionally will cause hooks to error out,
+   * you best assert their existence & throw otherwise.
+   * This missing would be a programming error that you should
+   * catch early anyways.
+   */
+  assert(splitApi.hooks.getPost, 'Endpoint `getPost` not loaded!');
+  const { data } = splitApi.hooks.getPost.useQuery(id);
```
Comment (msutkowski) [inline examples/react/src/features/bundleSplitting/Post.tsx:28]:
```suggestion
  const { data, error } = splitApi.hooks.getPost.useQuery(id);
```

**Review by msutkowski — CHANGES_REQUESTED** [review]

**Pull request merged**